liquid-compiler/src/grammar.pest

```diff
@@ -0,0 +1,56 @@
+WHITESPACE = _{" "}
+LINE_OR_SPACE = _{ NEWLINE | WHITESPACE }
+LiquidFile = ${ SOI ~ Element* ~ EOI }
+
+// Element-level parsing
+Element = _{ Expression | Tag | Raw } 
+
+TagStart = _{ (LINE_OR_SPACE* ~ "{%-") | "{%" }
+TagEnd =   _{ ("-%}" ~ LINE_OR_SPACE*) | "%}" }
+ExpressionStart = _{ (LINE_OR_SPACE* ~ "{{-") | "{{" }
+ExpressionEnd =   _{ ("-}}" ~ LINE_OR_SPACE*) | "}}" }
+
+Tag = !{ TagStart ~ Identifier ~ TagToken* ~ TagEnd }
+Expression = !{ ExpressionStart ~ FilterChain ~ ExpressionEnd }
+// Not allowing Tag/Expression Start/End might become a problem 
+// for {% raw %} tag: unclosed delimiters won't be accepted
+Raw = @{ (!(TagStart | ExpressionStart | TagEnd | ExpressionEnd) ~ ANY)+ }              
```

Comment:
The downside to pest is trying to decipher it.
- I get that `~` is "followed by".
- I'm assuming `_` means non-addressable. 
- `@` I assume means addressable
- `!{`  from the docs seems to be never addressable.?

Reply:
- `~` is followed by.
- `_` means that it doesn't produce a `Pair`. So when 'unwrapping' (as in, calling into_inner() ) `Tag`, for example, I would get an iterator with `Identifier` followed by `TagToken`s.
- `@` means the rule is 'atomic' ( as in, it can't be subdivided into other rules). This means that instead of Raw being an iterator over `ANY` (means any character is accepted), it is the text as a whole. It won't be subdivided on an iterator over characters.
- `!something` means to look into the next token and make sure its not `something`. So Raw will be `ANY`thing as long as `ANY` is not, for example, part of a `{{`.

Reply:
> !something means to look into the next token and make sure its not something. So Raw will be ANYthing as long as ANY is not, for example, part of a {{.

I chose a bad example.  That one was obvious.  The confusion comes in when `!` is reused in other syntax like

> Tag = !{ TagStart ~ Identifier ~ TagToken* ~ TagEnd }

Reply:
Oh. Basicaly, when I do `LiquidFile = ${ SOI ~ Element* ~ EOI }`, the $ prevents implicit whitespace. This is needed because otherwise all spaces after closing an Expression/Tag would be trimmed istead of being parsed as Raw. But $ will disable whitespace in the inner rules too (in this case the whole file because we're wrapping `LiquidFile`  rule with the $).

The !, on the other hand, is the exception to `$`: no matter whether the outter rule has `$` or not, if the inner rule is marked with ! it will never prevent implicit whitespace. This is needed inside Tags and Expressions so the whitespaces are trimmed as expected.